### PR TITLE
Repair --default-language, and highlight using full filename

### PR DESCRIPTION
**Pull request by th1000s**
Repair --default-language, and highlight using full filename

Fixed that the "txt" fallback was used instead of --default-language

And when looking for syntax highlighting, keep the filename around
as long as possible.

Using simple custom logic (filename > 4) a Makefile can be highlight with
make syntax, but a cc or ini file does not get treated like a .cc or .ini file.

Currently the underlying highlighting lib syntect and the sublime syntex
definitions can not make this distinction
(http://www.sublimetext.com/docs/syntax.html).

----

Plus clippy warnings

### CODE REVIEW

**Comment by th1000s** [comment]
The filename cutoff it 4, not 3 (from actual DOS 8.3 filenames) because of newer extensions like `.toml` and `.flac`. And I could not think of file formats without extensions which are just 4 short and have a meaning by itself, like `Makefile`. Maybe this could even be extended to 5.

This should fix #1513

**Comment by th1000s** [comment]
The filename cutoff is 4, not 3 (from actual DOS 8.3 filenames) because of newer extensions like `.toml` and `.flac`. And I could not think of file formats without extensions which are just 4 short and have a meaning by itself, like `Makefile`. Maybe this could even be extended to 5.

This should fix #1513

File: src/handlers/diff_header.rs [hunk 1]
```diff
@@ -479,49 +472,48 @@ mod tests {
...
-                "--- src/one.rs	2019-11-20 06:47:56.000000000 +0100"
-            ),
-            Some("rs")
+            get_filename_from_marker_line("--- src/one.rs	2019-11-20 06:47:56.000000000 +0100"),
+            Some("one.rs")
         );
     }
 
     #[test]
     fn test_get_file_extension_from_diff_header_line() {
```
Comment (dandavison) [inline src/handlers/diff_header.rs:481]: Looks like this test name needs to be updated

File: src/handlers/grep.rs [hunk 2]
```diff
@@ -116,12 +116,8 @@ impl<'a> StateMachine<'a> {
...
                 );
                 if new_path {
-                    if let Some(lang) = handlers::diff_header::get_extension(&grep_line.path)
-                        .or(self.config.default_language.as_deref())
-                    {
-                        self.painter.set_syntax(Some(lang));
-                        self.painter.set_highlighter();
-                    }
+                    self.painter.set_syntax(Some(grep_line.path.as_ref()));
+                    self.painter.set_highlighter();
```
Comment (dandavison) [inline src/handlers/grep.rs:120]: That's nicer.

File: src/handlers/diff_header.rs [hunk 3]
```diff
@@ -479,49 +472,48 @@ mod tests {
...
-            Some("rs")
+            get_filename_from_diff_header_line_file_path("wat hello.rs"),
+            Some("wat hello.rs")
         );
         assert_eq!(
-            get_file_extension_from_diff_header_line_file_path("/dev/null"),
+            get_filename_from_diff_header_line_file_path("/dev/null"),
             None
         );
+        // TODO, handle this:
```
Comment (dandavison) [inline src/handlers/diff_header.rs:502]: Can you expand comment to clarify in what sense this is not handled?

**Review by dandavison — CHANGES_REQUESTED** [review]
Thanks @th1000s! Just that one question about the code comment in the test: in what sense is Makefile etc not handled.

**Comment by tbm** [comment]
@th1000s I just saw you submitted a PR for an issue I reported. Thank you! do you by any chance to incorporate the feedback comments from @dandavison ?  I'd love to see your fix applied. Thanks!

File: src/handlers/diff_header.rs (hunk 1 above)
Comment (th1000s) [reply src/handlers/diff_header.rs]: Done.

File: src/handlers/diff_header.rs (hunk 3 above)
Comment (th1000s) [reply src/handlers/diff_header.rs]: Oh, that is handled now :) - Left over from writing the failing tests first.

**Comment by dandavison** [comment]
Thanks @th1000s!

(The MacOS build failure occurs on `main`; I am not sure why yet)

**Pull request merged**